File: js/leetcode/graph/permutation.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const result = [];
  const prevElements = [];

  const dfs = (elements) => {
    if (elements.length === 0) result.push([...prevElements]);

    for (let element of elements) {
      const nextElements = elements.filter((v) => v !== element);

      prevElements.push(element);
      dfs(nextElements);
      prevElements.pop();
    }
  };

  dfs(nums);
  return result;
};

console.log(permute([1, 2, 3]));
console.log(permute([0, 1]));
console.log(permute([1]));

File: python/leetcode/graph/permutation.py
"""
leetcode permutation 20221214
"""
import itertools
from typing import List


class Solution:
    """
    :param {None}
    :return {None}
    """

    def permute(self, nums: List[int]) -> List[List[int]]:
        """
        :param {self, List[int]}
        :return {List[List[int]]}
        """
        return list(map(list, itertools.permutations(nums)))


print(Solution().permute([1, 2, 3]))
print(Solution().permute([0, 1]))
print(Solution().permute([1]))
